crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs

```diff
@@ -0,0 +1,204 @@
+/// ## Example
+///
+/// ```python
+/// test1 = "example"
+/// if test1:
+///     print(test1)
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// if test1 := "example":
+///     print(test1)
+/// ```
+///
+/// ## References
+/// - [PEP 572 – Assignment Expressions](https://peps.python.org/pep-0572/)
+/// - [What’s New In Python 3.8 - Assignment Expressions](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions)
+#[derive(ViolationMetadata)]
+pub(crate) struct UnnecessaryAssignment {
+    name: Name,
+    assignment: String,
+    parentheses: bool,
+}
+
+impl UnnecessaryAssignment {
+    fn get_fix(&self) -> String {
+        let UnnecessaryAssignment {
+            name,
+            assignment,
+            parentheses,
+        } = self;
+        if *parentheses {
+            format!("({name} := {assignment})")
+        } else {
+            format!("{name} := {assignment}")
+        }
+    }
+}
+
+impl Violation for UnnecessaryAssignment {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use walrus operator `{}`.", self.get_fix())
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(format!(
+            "Move variable assignment into if statement using walrus operator `{}`.",
+            self.get_fix()
+        ))
+    }
+}
+
+type AssignmentBeforeIfStmt<'a> = (&'a Expr, &'a ExprName, &'a StmtAssign);
+
+/// PLR6103
+pub(crate) fn unnecessary_assignment(checker: &mut Checker, stmt: &StmtIf) {
+    // early out - unsupported python versions
+    if checker.settings.target_version < PythonVersion::Py38 {
+        return;
+    }
+
+    let if_test = &*stmt.test;
+
+    let previous_assignment = PreviousAssignment::new(checker.semantic());
+    let mut diagnostics: Vec<Diagnostic> = Vec::new();
+
+    // early out - only simple if checks, without elif or else clauses
+    if !&stmt.elif_else_clauses.is_empty() {
+        return;
+    }
+
+    // case - simple if checks (`if test1:`)
+    if let Some(unreferenced_binding) =
+        find_assignment_before_if_stmt(&previous_assignment, if_test, if_test)
+    {
+        diagnostics.push(create_diagnostic(checker, unreferenced_binding));
+    };
+
+    match &*stmt.test {
+        // case - bool operations (`if test1 and test2:`)
+        Expr::BoolOp(expr) => diagnostics.extend(expr.values.iter().filter_map(|bool_test| {
+            Some(create_diagnostic(
+                checker,
+                find_assignment_before_if_stmt(&previous_assignment, if_test, bool_test)?,
+            ))
+        })),
+
+        // case - compare (`if test1 is not None:`)
+        Expr::Compare(compare) => {
+            if let Some(error) =
+                find_assignment_before_if_stmt(&previous_assignment, if_test, &compare.left)
+            {
+                diagnostics.push(create_diagnostic(checker, error));
+            };
+        }
+
+        _ => {}
+    }
```

Comment:
I am curious if we should actually do this part. If I look at the ecosystem checks, these also produce some possible controversial results. 🤔 

Without it, I think most of the more controversial results would disappear.